File: backend-go/cli/tui.go
package cli

import (
	"fmt"

	"github.com/AlecAivazis/survey/v2"

	"file-lite-go/config"
)

func RunTui(getCtx func() CmdCtx, shouldContinue func() bool) error {
	cmds := Commands()
	for shouldContinue() {
		ctx := getCtx()
		options := make([]string, 0, len(cmds))
		idByLabel := make(map[string]string, len(cmds))
		for _, cmd := range cmds {
			if cmd.When != nil && !cmd.When(ctx) {
				continue
			}
			options = append(options, cmd.Label)
			idByLabel[cmd.Label] = cmd.ID
		}

		var selected string
		err := survey.AskOne(&survey.Select{
			Message: fmt.Sprintf("%s v%s Select function", config.PkgName, config.Version),
			Options: options,
		}, &selected)
		if err != nil {
			return err
		}

		id := idByLabel[selected]
		for _, cmd := range cmds {
			if cmd.ID == id {
				if err := cmd.Run(getCtx()); err != nil {
					return err
				}
				break
			}
		}
	}
	return nil
}

File: backend/src/cli/tui.ts
import type { CmdCtx } from './commands'
import { PKG_NAME, VERSION } from '@frontend/enum/version.ts'
import enquirer from 'enquirer'
import { commands } from './commands'

export async function runTui(getCtx: () => CmdCtx, shouldContinue: () => boolean) {
  while (shouldContinue()) {
    const ctx = getCtx()
    const choices = commands
      .filter(cmd => cmd.when?.(ctx) ?? true)
      .map(cmd => ({ message: cmd.label, name: cmd.id }))

    const { selectedId }: { selectedId: string } = await enquirer.prompt([{
      type: 'select',
      name: 'selectedId',
      message: `${PKG_NAME} v${VERSION} Select function`,
      choices,
    }])

    const cmd = commands.find(item => item.id === selectedId)
    if (cmd) {
      await cmd.run(getCtx())
    }
  }
}
